src/components/drive-container.tsx

```diff
@@ -0,0 +1,136 @@
+import { useDocumentDrive } from 'src/hooks/useDocumentDrive';
+
+function mapDocumentDriveNodeToTreeItem(
+    node: Immutable<Node>,
+    allNodes: Immutable<Array<Node>>
+): TreeItem {
+    const isFolder = node.kind === 'folder';
+    const pathRegex = new RegExp(`^${node.path}/[^/]+`);
+
+    return {
+        id: node.path,
+        label: node.name,
+        type: isFolder ? ItemType.Folder : ItemType.File,
+        children: isFolder
+            ? allNodes
+                  .filter(childrenNode => pathRegex.test(childrenNode.path))
+                  .map(childrenNode =>
+                      mapDocumentDriveNodeToTreeItem(childrenNode, allNodes)
+                  )
+            : undefined,
+    };
+}
+
+function mapDocumentDriveToTreeItem(drive: Immutable<Drive>): DriveTreeItem {
+    return {
+        id: drive.id,
+        label: drive.name,
+        type: ItemType.LocalDrive,
+        expanded: true,
+        children: drive.nodes
+            .filter(node => !node.path.includes('/'))
+            .map(node => mapDocumentDriveNodeToTreeItem(node, drive.nodes)),
+    };
+}
+
+export default function () {
+    const [drives, setDrives] = useState<DriveTreeItem[]>([]);
+    const { documentDrive, openFile, addFolder, deleteNode, renameNode } =
+        useDocumentDrive();
+
+    useEffect(() => {
+        const newItems = documentDrive?.state.drives.map(
+            mapDocumentDriveToTreeItem
+        );
+        setDrives(newItems ?? []);
+    }, [documentDrive]);
+
+    function handleNodeClick(item: TreeItem, drive: DriveTreeItem) {
+        if (item.type === ItemType.File) {
+            openFile(drive.id, item.id);
+        }
+    }
+
+    function addNewFolder(item: TreeItem, drive: DriveTreeItem) {
+        const basePath = item.id.replace(drive.id, '');
+        addFolder(
+            drive.id,
+            `${basePath ? `${basePath}/` : ''}new-folder`, // TODO check if there is a new folder already
+            'New Folder'
+        );
+    }
+
+    function updateItem(item: TreeItem, drive: DriveTreeItem) {
+        setDrives(drives =>
+            drives.map(_drive => {
+                if (_drive.id !== drive.id) {
+                    return _drive;
+                }
+                return traverseTree(drive, _item =>
+                    _item.id !== item.id ? _item : item
+                );
+            })
+        );
+    }
+
+    function setItemUpdate(item: TreeItem, drive: DriveTreeItem) {
+        updateItem(
+            {
+                ...item,
+                action: ActionType.Update,
+            },
+            drive
+        );
+    }
+
+    async function updateNodeName(item: TreeItem, drive: DriveTreeItem) {
+        renameNode(drive.id, item.id, item.label);
+    }
+
+    const handleItemOptionsClick: OnItemOptionsClickHandler = (
+        item,
+        option,
+        drive
+    ) => {
+        switch (option) {
+            case 'new-folder':
+                addNewFolder(item, drive);
```

Comment:
Here I think we should create a temp node in the tree (before creating the folder in the document) and set the property `action` to `ActionType.New`.

This will create a "virtual" folder that the user can rename immediately. After the user confirms or rejects the rename operation, the `onSubmitInput` callback will be executed with the item created before (you can check the property `action` to know if the user is trying to create a new folder or renaming an existing one). In this step you should create the folder in the document.

Reference:
https://github.com/powerhouse-inc/design-system/blob/main/src/connect/components/tree-view/tree-view.stories.tsx#L148-L166

https://github.com/powerhouse-inc/design-system/blob/main/src/connect/components/tree-view/tree-view.stories.tsx#L148-L166

Reply:
I've added this implementation in this [PR](https://github.com/powerhouse-inc/document-model-electron/pull/10)

Reply:
Ah yes, that is better.